In [1]:
#Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.metrics import confusion_matrix, precision_recall_fscore_support, accuracy_score

In [2]:
#Reading bot the train and test dataset
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

# Exploratory Data Analysis begins here

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252 entries, 0 to 251
Data columns (total 30 columns):
Game ID                                                             252 non-null int64
Team 1                                                              252 non-null object
Team 2                                                              252 non-null object
City                                                                252 non-null object
DayOfWeek                                                           252 non-null int64
DateOfGame                                                          252 non-null object
TimeOfGame                                                          252 non-null object
AvgWindSpeed                                                        252 non-null int64
AvgHumidity                                                         252 non-null float64
Inn 1 Team 1 NOP R>25,SR>125                                        252 non-null int64
Inn 1 Team 1 NOP R

In [4]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 30 columns):
Game ID                                                             76 non-null int64
Team 1                                                              76 non-null object
Team 2                                                              76 non-null object
CityOfGame                                                          76 non-null object
Day                                                                 76 non-null int64
DateOfGame                                                          76 non-null object
TimeOfGame                                                          76 non-null object
AvgWindSpeed                                                        76 non-null int64
AvgHumidity                                                         76 non-null float64
Inn 1 Team 1 NOP R>25,SR>125                                        76 non-null int64
Inn 1 Team 1 NOP R<25, SR>125 

In [5]:
train.head()

Game ID                        Team 1                       Team 2  \
0        1   Koramangala Traffic Jammers    Whitefield Water Loggers    
1        2  Electronic City Power Savers        Silkboard Slow Movers   
2        3       Indranagar Pub Watchers       Sarjapur Water Tankers   
3        4      Bellandur Froth Fighters  Koramangala Traffic Jammers   
4        5           Marathalli Chokers     Whitefield Water Loggers    

          City  DayOfWeek  DateOfGame TimeOfGame  AvgWindSpeed  AvgHumidity  \
0   Whitefield          1  01-01-2012   20:00:00             6         0.49   
1    Silkboard          2  01-02-2012   17:00:00             7         0.44   
2     Sarjapur          3  01-03-2012   20:30:00            11         0.23   
3  Koramangala          4  01-04-2012   16:00:00             6         0.61   
4   Marathalli          5  01-05-2012   20:00:00             6         0.56   

   Inn 1 Team 1 NOP R>25,SR>125  ...  Inn 2 Team 2 NOP R<25, SR>125  \
0                             1  ...                              0   
1                             3  ...                              2   
2                             2  ...                              1   
3                             0  ...                              0   
4                             3  ...                              3   

   Inn 2 Team 2 Total 4s  Inn 2 Team 2 Total 6s  \
0                      3                      3   
1                     18                      9   
2                     18                      1   
3                      5                      4   
4                     15                      6   

   Inn 2 Team 2 Max Strike Rate_ALLBatsmen  Inn 2 Team 1 NoP fast bowlers  \
0                                   120.00                              5   
1                                   215.15                              4   
2                                   300.00                              3   
3                                   100.00                              4   
4                                   205.26                              4   

   Inn 2 Team 1 NoP Spinners  Inn 2 Team 1 wickets taken_catches_runout  \
0                          0                                          6   
1                          1                                          4   
2                          3                                          0   
3                          2                                          3   
4                          2                                          4   

   Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping  \
0                                                  4                  
1                                                  0                  
2                                                  1                  
3                                                  2                  
4                                                  1                  

   Inn 2 Team 1 Extras conceded in_wides_No Balls  Winner (team 1=1, team 2=0)  
0                                              11                            1  
1                                               5                            1  
2                                              10                            0  
3                                              16                            0  
4                                               5                            0  

[5 rows x 30 columns]

In [6]:
test.head()

Game ID                        Team 1                  Team 2  \
0      253  Electronic City Power Savers     Marathalli Chokers    
1      254   Koramangala Traffic Jammers  Sarjapur Water Tankers   
2      255          HSR High Rent Payers     Marathalli Chokers    
3      256       Indranagar Pub Watchers   Silkboard Slow Movers   
4      257     Whitefield Water Loggers   Sarjapur Water Tankers   

        CityOfGame  Day  DateOfGame TimeOfGame  AvgWindSpeed  AvgHumidity  \
0  Electronic City    6  01-01-2016   20:00:00             5         0.62   
1      Koramangala    7  01-02-2016   20:00:00             6         0.66   
2       Marathalli    1  01-03-2016   16:00:00             5         0.64   
3       Indranagar    2  01-04-2016   20:00:00             5         0.64   
4       Whitefield    3  01-05-2016   16:00:00             5         0.62   

   Inn 1 Team 1 NOP R>25,SR>125  ...  Inn 2 Team 2 NOP R<25, SR>125  \
0                             1  ...                              0   
1                             1  ...                              1   
2                             0  ...                              1   
3                             1  ...                              3   
4                             2  ...                              2   

   Inn 2 Team 2 Total 4s  Inn 2 Team 2 Total 6s  \
0                      9                      5   
1                      8                      4   
2                      8                      1   
3                     14                      4   
4                     13                      2   

   Inn 2 Team 2 Max Strike Rate_ALLBatsmen  Inn 2 Team 1 NoP fast bowlers  \
0                                   142.85                              3   
1                                   210.00                              3   
2                                   166.66                              4   
3                                   166.66                              2   
4                                   160.00                              4   

   Inn 2 Team 1 NoP Spinners  Inn 2 Team 1 wickets taken_catches_runout  \
0                          3                                          1   
1                          2                                          1   
2                          2                                          2   
3                          4                                          6   
4                          2                                          6   

   Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping  \
0                                                  1                  
1                                                  1                  
2                                                  7                  
3                                                  3                  
4                                                  1                  

   Inn 2 Team 1 Extras conceded in_wides_No Balls  Winner (team 1=1, team 2=0)  
0                                               3                            0  
1                                               1                            0  
2                                               4                            1  
3                                               2                            1  
4                                               2                            1  

[5 rows x 30 columns]

In [7]:
train.describe()

Game ID   DayOfWeek  AvgWindSpeed  AvgHumidity  \
count  252.000000  252.000000    252.000000   252.000000   
mean   126.500000    4.011905      7.440476     0.625198   
std     72.890329    2.000960      2.353354     0.131620   
min      1.000000    1.000000      3.000000     0.160000   
25%     63.750000    2.000000      6.000000     0.560000   
50%    126.500000    4.000000      7.000000     0.655000   
75%    189.250000    6.000000      8.000000     0.720000   
max    252.000000    7.000000     22.000000     0.880000   

       Inn 1 Team 1 NOP R>25,SR>125  Inn 1 Team 1 NOP R<25, SR>125  \
count                     252.00000                     252.000000   
mean                        1.43254                       1.531746   
std                         0.98566                       1.185480   
min                         0.00000                       0.000000   
25%                         1.00000                       1.000000   
50%                         1.00000                       1.000000   
75%                         2.00000                       2.000000   
max                         4.00000                       6.000000   

       Inn 1 Team 1 Total 4s  Inn 1 Team 1 Total 6s  \
count             252.000000             252.000000   
mean               13.726190               4.880952   
std                 4.666618               2.996963   
min                 0.000000               0.000000   
25%                11.000000               3.000000   
50%                14.000000               5.000000   
75%                17.000000               7.000000   
max                25.000000              17.000000   

       Inn 1 Team 1 Max Strike Rate_ALLBatsmen  Inn 1 Team 2 NoP fast bowlers  \
count                               252.000000                     252.000000   
mean                                198.511627                       3.825397   
std                                  69.901951                       1.006587   
min                                   0.000000                       0.000000   
25%                                 161.110000                       3.000000   
50%                                 185.710000                       4.000000   
75%                                 220.000000                       4.000000   
max                                 600.000000                       6.000000   

       ...  Inn 2 Team 2 NOP R<25, SR>125  Inn 2 Team 2 Total 4s  \
count  ...                     252.000000             252.000000   
mean   ...                       1.250000              12.658730   
std    ...                       1.095536               4.379712   
min    ...                       0.000000               0.000000   
25%    ...                       0.000000              10.000000   
50%    ...                       1.000000              12.000000   
75%    ...                       2.000000              16.000000   
max    ...                       5.000000              25.000000   

       Inn 2 Team 2 Total 6s  Inn 2 Team 2 Max Strike Rate_ALLBatsmen  \
count             252.000000                               252.000000   
mean                4.519841                               188.509246   
std                 2.753412                                56.843114   
min                 0.000000                                 0.000000   
25%                 2.750000                               153.092500   
50%                 4.000000                               185.710000   
75%                 6.000000                               222.220000   
max                14.000000                               385.710000   

       Inn 2 Team 1 NoP fast bowlers  Inn 2 Team 1 NoP Spinners  \
count                     252.000000                 252.000000   
mean                        3.674603                   2.146825   
std                         1.000632                   0.976961   
min                         0.000000                   0.000000   
25%   

In [8]:
test.describe()

Game ID        Day  AvgWindSpeed  AvgHumidity  \
count   76.000000  76.000000     76.000000    76.000000   
mean   290.500000   3.986842      6.842105     0.660263   
std     22.083176   2.023155      1.405253     0.073266   
min    253.000000   1.000000      5.000000     0.500000   
25%    271.750000   2.000000      6.000000     0.600000   
50%    290.500000   4.000000      7.000000     0.660000   
75%    309.250000   6.000000      8.000000     0.730000   
max    328.000000   7.000000     10.000000     0.750000   

       Inn 1 Team 1 NOP R>25,SR>125  Inn 1 Team 1 NOP R<25, SR>125  \
count                     76.000000                      76.000000   
mean                       1.342105                       1.355263   
std                        0.973689                       1.162801   
min                        0.000000                       0.000000   
25%                        1.000000                       1.000000   
50%                        1.000000                       1.000000   
75%                        2.000000                       2.000000   
max                        3.000000                       6.000000   

       Inn 1 Team 1 Total 4s  Inn 1 Team 1 Total 6s  \
count              76.000000              76.000000   
mean               12.907895               5.157895   
std                 4.910336               3.072687   
min                 0.000000               0.000000   
25%                10.000000               3.000000   
50%                13.000000               5.000000   
75%                16.000000               7.000000   
max                24.000000              14.000000   

       Inn 1 Team 1 Max Strike Rate_ALLBatsmen  Inn 1 Team 2 NoP fast bowlers  \
count                                76.000000                      76.000000   
mean                                202.138947                       3.657895   
std                                  81.633354                       1.027004   
min                                   0.000000                       0.000000   
25%                                 157.480000                       3.000000   
50%                                 182.005000                       4.000000   
75%                                 233.925000                       4.000000   
max                                 600.000000                       6.000000   

       ...  Inn 2 Team 2 NOP R<25, SR>125  Inn 2 Team 2 Total 4s  \
count  ...                      76.000000              76.000000   
mean   ...                       1.421053              12.236842   
std    ...                       1.224888               4.162911   
min    ...                       0.000000               0.000000   
25%    ...                       0.000000              10.000000   
50%    ...                       1.000000              12.000000   
75%    ...                       2.000000              14.250000   
max    ...                       5.000000              22.000000   

       Inn 2 Team 2 Total 6s  Inn 2 Team 2 Max Strike Rate_ALLBatsmen  \
count              76.000000                                76.000000   
mean                4.460526                               191.971974   
std                 2.918633                                72.436529   
min                 0.000000                                 0.000000   
25%                 2.000000                               154.012500   
50%                 4.000000                               175.000000   
75%                 6.000000                               219.312500   
max                14.000000                               400.000000   

       Inn 2 Team 1 NoP fast bowlers  Inn 2 Team 1 NoP Spinners  \
count                      76.000000                  76.000000   
mean                        3.328947                   2.223684   
std                         1.135936                   1.078254   
min                         0.000000                   0.000000   
25%            

## Initial Analysis of Data

__Train Data__: 252 data points, 30 columns

__Test Data__: 76 data points, 30 columns

Also, the featues __Team 1__, __Team 2__, __City__, __TimeOfGame__ and __DayOfWeek__ are categorical features.

Others are numeric features and our predictor variable is __Winner (team 1=1, team 2=0)__

## Next, we check for any missing values.

In [9]:
train.isnull().sum()

Game ID                                                             0
Team 1                                                              0
Team 2                                                              0
City                                                                0
DayOfWeek                                                           0
DateOfGame                                                          0
TimeOfGame                                                          0
AvgWindSpeed                                                        0
AvgHumidity                                                         0
Inn 1 Team 1 NOP R>25,SR>125                                        0
Inn 1 Team 1 NOP R<25, SR>125                                       0
Inn 1 Team 1 Total 4s                                               0
Inn 1 Team 1 Total 6s                                               0
Inn 1 Team 1 Max Strike Rate_ALLBatsmen                             0
Inn 1 Team 2 NoP fas

In [10]:
test.isnull().sum()

Game ID                                                             0
Team 1                                                              0
Team 2                                                              0
CityOfGame                                                          0
Day                                                                 0
DateOfGame                                                          0
TimeOfGame                                                          0
AvgWindSpeed                                                        0
AvgHumidity                                                         0
Inn 1 Team 1 NOP R>25,SR>125                                        0
Inn 1 Team 1 NOP R<25, SR>125                                       0
Inn 1 Team 1 Total 4s                                               0
Inn 1 Team 1 Total 6s                                               0
Inn 1 Team 1 Max Strike Rate_ALLBatsmen                             0
Inn 1 Team 2 NoP fas

__None of the columns have any missing values both in the train and test datasets__

### Renaming columns in Test set which have different name in the Train set but serve the same purpose

In [11]:
test = test.rename(columns = {"CityOfGame": "City", "Day":"DayOfWeek"})

### Dropping columns __Game ID__ and __DateOfGame__ as they do not add any information

In [12]:
train.drop("Game ID", axis=1, inplace=True)

In [13]:
test.drop("Game ID", axis=1, inplace=True)

In [14]:
train.drop("DateOfGame", axis=1, inplace=True)

In [15]:
test.drop("DateOfGame", axis=1, inplace=True)

## Standardizing Numeric Data

We use StandardScaler in order to standardize the numeric data and center it around the respective column mean.

Parameters mu and sigma are calculated from the training data and applied to both train and test data sets.

In [16]:
scaler = StandardScaler()

In [17]:
train[['AvgWindSpeed', 'AvgHumidity', 'Inn 1 Team 1 NOP R>25,SR>125', 'Inn 1 Team 1 NOP R<25, SR>125', 'Inn 1 Team 1 Total 4s',
       'Inn 1 Team 1 Total 6s', 'Inn 1 Team 1 Max Strike Rate_ALLBatsmen', 'Inn 1 Team 2 NoP fast bowlers', 'Inn 1 Team 2 NoP Spinners',
       'Inn 1 Team 2 wickets taken_catches_runout', 'Inn1 Team 2 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 1 Team 2 Extras conceded in_wides_No Balls', 'Inn 2 Team 2 NOP R>25,SR>125', 'Inn 2 Team 2 NOP R<25, SR>125',
       'Inn 2 Team 2 Total 4s', 'Inn 2 Team 2 Total 6s', 'Inn 2 Team 2 Max Strike Rate_ALLBatsmen', 'Inn 2 Team 1 NoP fast bowlers',
       'Inn 2 Team 1 NoP Spinners', 'Inn 2 Team 1 wickets taken_catches_runout', 'Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 2 Team 1 Extras conceded in_wides_No Balls']] = scaler.fit_transform(train[['AvgWindSpeed', 'AvgHumidity', 'Inn 1 Team 1 NOP R>25,SR>125', 'Inn 1 Team 1 NOP R<25, SR>125', 'Inn 1 Team 1 Total 4s',
       'Inn 1 Team 1 Total 6s', 'Inn 1 Team 1 Max Strike Rate_ALLBatsmen', 'Inn 1 Team 2 NoP fast bowlers', 'Inn 1 Team 2 NoP Spinners',
       'Inn 1 Team 2 wickets taken_catches_runout', 'Inn1 Team 2 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 1 Team 2 Extras conceded in_wides_No Balls', 'Inn 2 Team 2 NOP R>25,SR>125', 'Inn 2 Team 2 NOP R<25, SR>125',
       'Inn 2 Team 2 Total 4s', 'Inn 2 Team 2 Total 6s', 'Inn 2 Team 2 Max Strike Rate_ALLBatsmen', 'Inn 2 Team 1 NoP fast bowlers',
       'Inn 2 Team 1 NoP Spinners', 'Inn 2 Team 1 wickets taken_catches_runout', 'Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 2 Team 1 Extras conceded in_wides_No Balls']])

In [18]:
test[['AvgWindSpeed', 'AvgHumidity', 'Inn 1 Team 1 NOP R>25,SR>125', 'Inn 1 Team 1 NOP R<25, SR>125', 'Inn 1 Team 1 Total 4s',
       'Inn 1 Team 1 Total 6s', 'Inn 1 Team 1 Max Strike Rate_ALLBatsmen', 'Inn 1 Team 2 NoP fast bowlers', 'Inn 1 Team 2 NoP Spinners',
       'Inn 1 Team 2 wickets taken_catches_runout', 'Inn1 Team 2 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 1 Team 2 Extras conceded in_wides_No Balls', 'Inn 2 Team 2 NOP R>25,SR>125', 'Inn 2 Team 2 NOP R<25, SR>125',
       'Inn 2 Team 2 Total 4s', 'Inn 2 Team 2 Total 6s', 'Inn 2 Team 2 Max Strike Rate_ALLBatsmen', 'Inn 2 Team 1 NoP fast bowlers',
       'Inn 2 Team 1 NoP Spinners', 'Inn 2 Team 1 wickets taken_catches_runout', 'Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 2 Team 1 Extras conceded in_wides_No Balls']] = scaler.transform(test[['AvgWindSpeed', 'AvgHumidity', 'Inn 1 Team 1 NOP R>25,SR>125', 'Inn 1 Team 1 NOP R<25, SR>125', 'Inn 1 Team 1 Total 4s',
       'Inn 1 Team 1 Total 6s', 'Inn 1 Team 1 Max Strike Rate_ALLBatsmen', 'Inn 1 Team 2 NoP fast bowlers', 'Inn 1 Team 2 NoP Spinners',
       'Inn 1 Team 2 wickets taken_catches_runout', 'Inn1 Team 2 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 1 Team 2 Extras conceded in_wides_No Balls', 'Inn 2 Team 2 NOP R>25,SR>125', 'Inn 2 Team 2 NOP R<25, SR>125',
       'Inn 2 Team 2 Total 4s', 'Inn 2 Team 2 Total 6s', 'Inn 2 Team 2 Max Strike Rate_ALLBatsmen', 'Inn 2 Team 1 NoP fast bowlers',
       'Inn 2 Team 1 NoP Spinners', 'Inn 2 Team 1 wickets taken_catches_runout', 'Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping',
       'Inn 2 Team 1 Extras conceded in_wides_No Balls']])

## Train and Test Data after Numerical Attributes are Standardized

In [19]:
train

Team 1                        Team 2  \
0     Koramangala Traffic Jammers     Whitefield Water Loggers    
1    Electronic City Power Savers         Silkboard Slow Movers   
2         Indranagar Pub Watchers        Sarjapur Water Tankers   
3        Bellandur Froth Fighters   Koramangala Traffic Jammers   
4             Marathalli Chokers      Whitefield Water Loggers    
..                            ...                           ...   
247   Koramangala Traffic Jammers           Marathalli Chokers    
248     Whitefield Water Loggers   Electronic City Power Savers   
249   Koramangala Traffic Jammers           Marathalli Chokers    
250     Whitefield Water Loggers            Marathalli Chokers    
251  Electronic City Power Savers     Whitefield Water Loggers    

                City  DayOfWeek TimeOfGame  AvgWindSpeed  AvgHumidity  \
0         Whitefield          1   20:00:00     -0.613313    -1.029230   
1          Silkboard          2   17:00:00     -0.187542    -1.409867   
2           Sarjapur          3   20:30:00      1.515542    -3.008543   
3        Koramangala          4   16:00:00     -0.613313    -0.115702   
4         Marathalli          5   20:00:00     -0.613313    -0.496339   
..               ...        ...        ...           ...          ...   
247      Koramangala          4   20:00:00      0.664000     1.102337   
248       Marathalli          5   20:00:00     -0.613313     0.645572   
249       Marathalli          6   20:00:00     -0.187542     0.721700   
250  Electronic City          7   20:00:00      0.238229     0.645572   
251  Electronic City          1   20:00:00      0.238229     0.569445   

     Inn 1 Team 1 NOP R>25,SR>125  Inn 1 Team 1 NOP R<25, SR>125  \
0                       -0.439706                      -0.449442   
1                        1.593429                       0.395777   
2                        0.576862                       0.395777   
3                       -1.456273                       0.395777   
4                        1.593429                      -0.449442   
..                            ...                            ...   
247                      1.593429                       0.395777   
248                      0.576862                      -0.449442   
249                      0.576862                      -1.294661   
250                      0.576862                      -0.449442   
251                      0.576862                       2.086215   

     Inn 1 Team 1 Total 4s  ...  Inn 2 Team 2 NOP R<25, SR>125  \
0                 0.273505  ...                      -1.143265   
1                 1.347077  ...                       0.685959   
2                -0.155924  ...                      -0.228653   
3                -1.658924  ...                      -1.143265   
4                 0.917649  ...                       1.600570   
..                     ...  ...                            ...   
247               0.488220  ...                      -0.228653   
248              -0.155924  ...                      -1.143265   
249               0.273505  ...                       0.685959   
250              -0.155924  ...                       1.600570   
251              -1.444210  ...                       1.600570   

     Inn 2 Team 2 Total 4s  Inn 2 Team 2 Total 6s  \
0                -2.209723              -0.553083   
1                 1.221975               1.630368   
2                 1.221975              -1.280900   
3                -1.752164              -0.189175   
4                 0.535635               0.538642   
..                     ...                    ...   
247               1.679535              -0.553083   
248               0.306856               1.994276   
249               1.221975              -0.553083   
250               0.078076              -0.189175   
251              -0.379484              -0.189175   

     Inn 2 Team 2 Max Strike Rate_ALLBatsmen  Inn 2 Team 1 NoP fast bowlers  \
0                  

In [20]:
test

Team 1                       Team 2  \
0   Electronic City Power Savers          Marathalli Chokers    
1    Koramangala Traffic Jammers       Sarjapur Water Tankers   
2           HSR High Rent Payers          Marathalli Chokers    
3        Indranagar Pub Watchers        Silkboard Slow Movers   
4      Whitefield Water Loggers        Sarjapur Water Tankers   
..                           ...                          ...   
71       Indranagar Pub Watchers          Marathalli Chokers    
72   Koramangala Traffic Jammers       Sarjapur Water Tankers   
73  Electronic City Power Savers          Marathalli Chokers    
74  Electronic City Power Savers       Sarjapur Water Tankers   
75  Electronic City Power Savers  Koramangala Traffic Jammers   

               City  DayOfWeek TimeOfGame  AvgWindSpeed  AvgHumidity  \
0   Electronic City          6   20:00:00     -1.039084    -0.039574   
1       Koramangala          7   20:00:00     -0.613313     0.264935   
2        Marathalli          1   16:00:00     -1.039084     0.112681   
3        Indranagar          2   20:00:00     -1.039084     0.112681   
4        Whitefield          3   16:00:00     -1.039084    -0.039574   
..              ...        ...        ...           ...          ...   
71       Indranagar          7   20:00:00     -0.187542     0.950082   
72              HSR          1   20:00:00     -0.613313     0.873955   
73       Whitefield          2   20:00:00     -0.613313     0.873955   
74  Electronic City          3   20:00:00     -0.613313     0.873955   
75  Electronic City          4   20:00:00     -0.187542     0.950082   

    Inn 1 Team 1 NOP R>25,SR>125  Inn 1 Team 1 NOP R<25, SR>125  \
0                      -0.439706                      -0.449442   
1                      -0.439706                       1.240996   
2                      -1.456273                       0.395777   
3                      -0.439706                       1.240996   
4                       0.576862                      -1.294661   
..                           ...                            ...   
71                     -0.439706                       0.395777   
72                      0.576862                      -0.449442   
73                      1.593429                      -1.294661   
74                      1.593429                      -0.449442   
75                      1.593429                      -0.449442   

    Inn 1 Team 1 Total 4s  ...  Inn 2 Team 2 NOP R<25, SR>125  \
0               -1.658924  ...                      -1.143265   
1               -1.658924  ...                      -0.228653   
2               -0.585352  ...                      -0.228653   
3                1.991221  ...                       1.600570   
4                0.058791  ...                       0.685959   
..                    ...  ...                            ...   
71              -0.155924  ...                      -1.143265   
72               0.058791  ...                       2.515182   
73               1.991221  ...                       1.600570   
74               2.205935  ...                       3.429794   
75              -0.155924  ...                       0.685959   

    Inn 2 Team 2 Total 4s  Inn 2 Team 2 Total 6s  \
0               -0.837044               0.174734   
1               -1.065824              -0.189175   
2               -1.065824              -1.280900   
3                0.306856              -0.189175   
4                0.078076              -0.916992   
..                    ...                    ...   
71               0.764415              -0.553083   
72               0.306856              -0.553083   
73               0.306856               0.174734   
74              -1.294604               0.902551   
75               1.221975               0.538642   

    Inn 2 Team 2 Max Strike Rate_ALLBatsmen  Inn 2 Team 1 NoP fast bowlers  \
0                                 -0.804849                      -0.675519   
1            

## Working with Categorical Data

### Initially we identify unique values for specific categorial attributes from both train and test data sets

__Categorical Attribute 1__ : Team Names (Team 1 and Team 2)

In [21]:
print(test["Team 1"].nunique(), "Teams in Test data:", test["Team 1"].unique())

9 Teams in Test data: ['Electronic City Power Savers' 'Koramangala Traffic Jammers'
 'HSR High Rent Payers' 'Indranagar Pub Watchers'
 'Whitefield Water Loggers ' 'Bellandur Froth Fighters'
 'Marathalli Chokers ' 'Silkboard Slow Movers' 'Sarjapur Water Tankers']


In [22]:
print(train["Team 1"].nunique(), "Teams in Test data:", train["Team 1"].unique())

10 Teams in Test data: ['Koramangala Traffic Jammers' 'Electronic City Power Savers'
 'Indranagar Pub Watchers' 'Bellandur Froth Fighters'
 'Marathalli Chokers ' 'Silkboard Slow Movers' 'Whitefield Water Loggers '
 'Sarjapur Water Tankers' 'Airport Flyers' 'Forum Fans']


In [23]:
unique_teams = list(set(train[['Team 1', 'Team 2']].as_matrix().reshape((1,-1)).tolist()[0]))

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [24]:
unique_teams = unique_teams + (list(set(test[['Team 1', 'Team 2']].as_matrix().reshape((1,-1)).tolist()[0])))

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [25]:
set(unique_teams)

{'Airport Flyers',
 'Bellandur Froth Fighters',
 'Electronic City Power Savers',
 'Forum Fans',
 'HSR High Rent Payers',
 'Indranagar Pub Watchers',
 'Koramangala Traffic Jammers',
 'Marathalli Chokers ',
 'Sarjapur Water Tankers',
 'Silkboard Slow Movers',
 'Whitefield Water Loggers '}

In [26]:
len(set(unique_teams))

11

## Conclusions drawn from __Categorical Attribute 1__

Train set consists of 10 unique team names, whereas test set contains 9 unique team names.

However, 1 value __HSR High Rent Payers__ appears only in the test data.

Ideally, this is an issue with the dataset because if we don't provide information about the categorical value, __HSR High Rent Payers__ to our model during training, then we shouldn't penalize our model for not being able to work properly with this value at test time.

Hence, for this presentation we would work with unique values from both the train and test data. Ideally our test data should be entirely unseen. However, in this presentation we will have an entry for __HSR High Rent Payers__ when creating a One Hot Encoding for the Team's attributes during training time itself even though the value appears only in the test data.

__Categorical Attribute 2__ : City

In [27]:
print(test["City"].nunique(), "Cities in Test data:", test["City"].unique())

9 Cities in Test data: ['Electronic City' 'Koramangala' 'Marathalli' 'Indranagar' 'Whitefield'
 'Bellandur' 'HSR' 'Sarjapur' 'Silkboard']


In [28]:
print(train["City"].nunique(), "Cities in Train data:", train["City"].unique())

8 Cities in Train data: ['Whitefield' 'Silkboard' 'Sarjapur' 'Koramangala' 'Marathalli'
 'Indranagar' 'Bellandur' 'Electronic City']


In [29]:
unique_cities = (list(set(test[['City']].as_matrix().reshape((1,-1)).tolist()[0])))

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [30]:
unique_cities = unique_cities + (list(set(train[['City']].as_matrix().reshape((1,-1)).tolist()[0])))

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [31]:
set(unique_cities)

{'Bellandur',
 'Electronic City',
 'HSR',
 'Indranagar',
 'Koramangala',
 'Marathalli',
 'Sarjapur',
 'Silkboard',
 'Whitefield'}

In [32]:
len(set(unique_cities))

9

## Conclusions drawn from __Categorical Attribute 2__

Train set consists of 8 unique city names, whereas test set contains 9 unique city names.

However, 1 value __HSR__ appears only in the test data.

Similar to __Categorical Attribute 1__, we will have an entry for __HSR__ when creating a One Hot Encoding for the City's attributes during training time itself even though the value appears only in the test data.

__Categorical Attribute 3__ : Time Of Game

In [33]:
print(train["TimeOfGame"].nunique(), "Timings in Train data:", train["TimeOfGame"].unique())

11 Timings in Train data: ['20:00:00' '17:00:00' '20:30:00' '16:00:00' '20:15:00' '12:30:00'
 '17:10:00' '18:50:00' '16:30:00' '16:45:00' '15:00:00']


In [34]:
print(test["TimeOfGame"].nunique(), "Timings in test data:", test["TimeOfGame"].unique())

2 Timings in test data: ['20:00:00' '16:00:00']


## Conclusions drawn from __Categorical Attribute 3__

Train set contains 11 unique timings but test contains only 2 unique timings. We can get rid of this attribute, however in this presentation we wish to keep all the attributes.

__Categorical Attribute 4__: Day of Week

This doesn't require much analysis as it holds 7 values as per the days of the week.

In [35]:
'''
We can use sort of a Lable Encoder to represent our Categorical Attribute values if we use a One Hot Encoding setting.
However, here we'll use the dummies feature from pandas and hence don't need this setting.
'''
# encode = {'Team 1': {'Whitefield Water Loggers ':1,'Silkboard Slow Movers':2,'Sarjapur Water Tankers':3,'Koramangala Traffic Jammers':4,'Indranagar Pub Watchers':5,'Marathalli Chokers ':6,'Electronic City Power Savers':7,'Bellandur Froth Fighters':8,'Forum Fans':9,'Airport Flyers':10,'HSR High Rent Payers':11},
#           'Team 2': {'Whitefield Water Loggers ':1,'Silkboard Slow Movers':2,'Sarjapur Water Tankers':3,'Koramangala Traffic Jammers':4,'Indranagar Pub Watchers':5,'Marathalli Chokers ':6,'Electronic City Power Savers':7,'Bellandur Froth Fighters':8,'Forum Fans':9,'Airport Flyers':10,'HSR High Rent Payers':11},
#           'City': {'Silkboard':1, 'Sarjapur':2, 'Koramangala':3, 'Marathalli':4, 'Indranagar':5, 'Bellandur':6, 'Electronic City':7, 'Whitefield':8, 'HSR':9}}
          

"\nWe can use sort of a Lable Encoder to represent our Categorical Attribute values if we use a One Hot Encoding setting.\nHowever, here we'll use the dummies feature from pandas and hence don't need this setting.\n"

In [36]:
#train.replace(encode, inplace=True)

In [37]:
#test.replace(encode, inplace=True)

We will be combining training and train data for One Hot Encoding purpose and in order to identify the train and test, we add a dummy column as represented in the next block. We will drop this later on.

In [38]:
train['train'] = 1
test['train'] = 0

In [39]:
combined = pd.concat([train, test])

In [40]:
combined

Team 1                       Team 2  \
0    Koramangala Traffic Jammers    Whitefield Water Loggers    
1   Electronic City Power Savers        Silkboard Slow Movers   
2        Indranagar Pub Watchers       Sarjapur Water Tankers   
3       Bellandur Froth Fighters  Koramangala Traffic Jammers   
4            Marathalli Chokers     Whitefield Water Loggers    
..                           ...                          ...   
71       Indranagar Pub Watchers          Marathalli Chokers    
72   Koramangala Traffic Jammers       Sarjapur Water Tankers   
73  Electronic City Power Savers          Marathalli Chokers    
74  Electronic City Power Savers       Sarjapur Water Tankers   
75  Electronic City Power Savers  Koramangala Traffic Jammers   

               City  DayOfWeek TimeOfGame  AvgWindSpeed  AvgHumidity  \
0        Whitefield          1   20:00:00     -0.613313    -1.029230   
1         Silkboard          2   17:00:00     -0.187542    -1.409867   
2          Sarjapur          3   20:30:00      1.515542    -3.008543   
3       Koramangala          4   16:00:00     -0.613313    -0.115702   
4        Marathalli          5   20:00:00     -0.613313    -0.496339   
..              ...        ...        ...           ...          ...   
71       Indranagar          7   20:00:00     -0.187542     0.950082   
72              HSR          1   20:00:00     -0.613313     0.873955   
73       Whitefield          2   20:00:00     -0.613313     0.873955   
74  Electronic City          3   20:00:00     -0.613313     0.873955   
75  Electronic City          4   20:00:00     -0.187542     0.950082   

    Inn 1 Team 1 NOP R>25,SR>125  Inn 1 Team 1 NOP R<25, SR>125  \
0                      -0.439706                      -0.449442   
1                       1.593429                       0.395777   
2                       0.576862                       0.395777   
3                      -1.456273                       0.395777   
4                       1.593429                      -0.449442   
..                           ...                            ...   
71                     -0.439706                       0.395777   
72                      0.576862                      -0.449442   
73                      1.593429                      -1.294661   
74                      1.593429                      -0.449442   
75                      1.593429                      -0.449442   

    Inn 1 Team 1 Total 4s  ...  Inn 2 Team 2 Total 4s  Inn 2 Team 2 Total 6s  \
0                0.273505  ...              -2.209723              -0.553083   
1                1.347077  ...               1.221975               1.630368   
2               -0.155924  ...               1.221975              -1.280900   
3               -1.658924  ...              -1.752164              -0.189175   
4                0.917649  ...               0.535635               0.538642   
..                    ...  ...                    ...                    ...   
71              -0.155924  ...               0.764415              -0.553083   
72               0.058791  ...               0.306856              -0.553083   
73               1.991221  ...               0.306856               0.174734   
74               2.205935  ...              -1.294604               0.902551   
75              -0.155924  ...               1.221975               0.538642   

    Inn 2 Team 2 Max Strike Rate_ALLBatsmen  Inn 2 Team 1 NoP fast bowlers  \
0                                 -1.207632                       1.327195   
1                                  0.469604                       0.325838   
2                                  1.965280                      -0.675519   
3                                 -1.560178                       0.325838   
4                                  0.295270                       0.325838   
..                                      ...                            ...   
71                                -0.678814                       0.

## Adding One Hot Encodings for the Categorical Features using Pandas' Get Dummies feature

In [41]:
dfTeam1 = pd.get_dummies(combined['Team 1'], prefix = 'Team 1')

In [42]:
dfTeam1

Team 1_Airport Flyers  Team 1_Bellandur Froth Fighters  \
0                       0                                0   
1                       0                                0   
2                       0                                0   
3                       0                                1   
4                       0                                0   
..                    ...                              ...   
71                      0                                0   
72                      0                                0   
73                      0                                0   
74                      0                                0   
75                      0                                0   

    Team 1_Electronic City Power Savers  Team 1_Forum Fans  \
0                                     0                  0   
1                                     1                  0   
2                                     0                  0   
3                                     0                  0   
4                                     0                  0   
..                                  ...                ...   
71                                    0                  0   
72                                    0                  0   
73                                    1                  0   
74                                    1                  0   
75                                    1                  0   

    Team 1_HSR High Rent Payers  Team 1_Indranagar Pub Watchers  \
0                             0                               0   
1                             0                               0   
2                             0                               1   
3                             0                               0   
4                             0                               0   
..                          ...                             ...   
71                            0                               1   
72                            0                               0   
73                            0                               0   
74                            0                               0   
75                            0                               0   

    Team 1_Koramangala Traffic Jammers  Team 1_Marathalli Chokers   \
0                                    1                           0   
1                                    0                           0   
2                                    0                           0   
3                                    0                           0   
4                                    0                           1   
..                                 ...                         ...   
71                                   0                           0   
72                                   1                           0   
73                                   0                           0   
74                                   0                           0   
75                                   0                           0   

    Team 1_Sarjapur Water Tankers  Team 1_Silkboard Slow Movers  \
0                               0                             0   
1                               0                             0   
2                               0                             0   
3                               0                             0   
4                               0                             0   
..                            ...                           ...   
71                              0                             0   
72                              0                             0   
73                              0                             0   
74                              0                             0   
75                              0                             0   

    Team 1_Whitefield Water Loggers   
0                       

In [43]:
dfTeam1.drop('Team 1_Airport Flyers', axis = 1, inplace = True)

In [44]:
dfTeam2 = pd.get_dummies(combined['Team 2'], prefix = 'Team 2', dummy_na=True)

In [45]:
dfTeam2

Team 2_Airport Flyers  Team 2_Bellandur Froth Fighters  \
0                       0                                0   
1                       0                                0   
2                       0                                0   
3                       0                                0   
4                       0                                0   
..                    ...                              ...   
71                      0                                0   
72                      0                                0   
73                      0                                0   
74                      0                                0   
75                      0                                0   

    Team 2_Electronic City Power Savers  Team 2_Forum Fans  \
0                                     0                  0   
1                                     0                  0   
2                                     0                  0   
3                                     0                  0   
4                                     0                  0   
..                                  ...                ...   
71                                    0                  0   
72                                    0                  0   
73                                    0                  0   
74                                    0                  0   
75                                    0                  0   

    Team 2_HSR High Rent Payers  Team 2_Indranagar Pub Watchers  \
0                             0                               0   
1                             0                               0   
2                             0                               0   
3                             0                               0   
4                             0                               0   
..                          ...                             ...   
71                            0                               0   
72                            0                               0   
73                            0                               0   
74                            0                               0   
75                            0                               0   

    Team 2_Koramangala Traffic Jammers  Team 2_Marathalli Chokers   \
0                                    0                           0   
1                                    0                           0   
2                                    0                           0   
3                                    1                           0   
4                                    0                           0   
..                                 ...                         ...   
71                                   0                           1   
72                                   0                           0   
73                                   0                           1   
74                                   0                           0   
75                                   1                           0   

    Team 2_Sarjapur Water Tankers  Team 2_Silkboard Slow Movers  \
0                               0                             0   
1                               0                             1   
2                               1                             0   
3                               0                             0   
4                               0                             0   
..                            ...                           ...   
71                              0                             0   
72                              1                             0   
73                              0                             0   
74                              1                             0   
75                              0                             0   

    Team 2_Whitefield Water Loggers   Team 2_nan  
0           

In [46]:
dfTeam2.drop('Team 2_Airport Flyers', axis = 1, inplace = True)

In [47]:
dfTimeOfGame = pd.get_dummies(combined['TimeOfGame'], prefix = 'TimeOfGame')

In [48]:
dfTimeOfGame

TimeOfGame_12:30:00  TimeOfGame_15:00:00  TimeOfGame_16:00:00  \
0                     0                    0                    0   
1                     0                    0                    0   
2                     0                    0                    0   
3                     0                    0                    1   
4                     0                    0                    0   
..                  ...                  ...                  ...   
71                    0                    0                    0   
72                    0                    0                    0   
73                    0                    0                    0   
74                    0                    0                    0   
75                    0                    0                    0   

    TimeOfGame_16:30:00  TimeOfGame_16:45:00  TimeOfGame_17:00:00  \
0                     0                    0                    0   
1                     0                    0                    1   
2                     0                    0                    0   
3                     0                    0                    0   
4                     0                    0                    0   
..                  ...                  ...                  ...   
71                    0                    0                    0   
72                    0                    0                    0   
73                    0                    0                    0   
74                    0                    0                    0   
75                    0                    0                    0   

    TimeOfGame_17:10:00  TimeOfGame_18:50:00  TimeOfGame_20:00:00  \
0                     0                    0                    1   
1                     0                    0                    0   
2                     0                    0                    0   
3                     0                    0                    0   
4                     0                    0                    1   
..                  ...                  ...                  ...   
71                    0                    0                    1   
72                    0                    0                    1   
73                    0                    0                    1   
74                    0                    0                    1   
75                    0                    0                    1   

    TimeOfGame_20:15:00  TimeOfGame_20:30:00  
0                     0                    0  
1                     0                    0  
2                     0                    1  
3                     0                    0  
4                     0                    0  
..                  ...                  ...  
71                    0                    0  
72                    0                    0  
73                    0                    0  
74                    0                    0  
75                    0                    0  

[328 rows x 11 columns]

In [49]:
dfCity = pd.get_dummies(combined['City'], prefix = 'City')

In [50]:
dfCity

City_Bellandur  City_Electronic City  City_HSR  City_Indranagar  \
0                0                     0         0                0   
1                0                     0         0                0   
2                0                     0         0                0   
3                0                     0         0                0   
4                0                     0         0                0   
..             ...                   ...       ...              ...   
71               0                     0         0                1   
72               0                     0         1                0   
73               0                     0         0                0   
74               0                     1         0                0   
75               0                     1         0                0   

    City_Koramangala  City_Marathalli  City_Sarjapur  City_Silkboard  \
0                  0                0              0               0   
1                  0                0              0               1   
2                  0                0              1               0   
3                  1                0              0               0   
4                  0                1              0               0   
..               ...              ...            ...             ...   
71                 0                0              0               0   
72                 0                0              0               0   
73                 0                0              0               0   
74                 0                0              0               0   
75                 0                0              0               0   

    City_Whitefield  
0                 1  
1                 0  
2                 0  
3                 0  
4                 0  
..              ...  
71                0  
72                0  
73                1  
74                0  
75                0  

[328 rows x 9 columns]

In [51]:
dfCity.drop('City_Bellandur', axis = 1, inplace = True)

In [52]:
dfDayOfWeek = pd.get_dummies(combined['DayOfWeek'], prefix = 'DayOfWeek')

In [53]:
dfDayOfWeek

DayOfWeek_1  DayOfWeek_2  DayOfWeek_3  DayOfWeek_4  DayOfWeek_5  \
0             1            0            0            0            0   
1             0            1            0            0            0   
2             0            0            1            0            0   
3             0            0            0            1            0   
4             0            0            0            0            1   
..          ...          ...          ...          ...          ...   
71            0            0            0            0            0   
72            1            0            0            0            0   
73            0            1            0            0            0   
74            0            0            1            0            0   
75            0            0            0            1            0   

    DayOfWeek_6  DayOfWeek_7  
0             0            0  
1             0            0  
2             0            0  
3             0            0  
4             0            0  
..          ...          ...  
71            0            1  
72            0            0  
73            0            0  
74            0            0  
75            0            0  

[328 rows x 7 columns]

In [54]:
dfDayOfWeek.drop('DayOfWeek_1', axis = 1, inplace = True)

In [55]:
# combined = pd.concat([combined, dfTeam1, dfTeam2, dfCity, dfDayOfWeek], axis=1)
combined = pd.concat([combined, dfTeam1, dfTeam2, dfCity, dfDayOfWeek, dfTimeOfGame], axis=1)

In [56]:
# combined = combined.drop(['Team 1', 'Team 2', 'City', 'DayOfWeek'], axis=1)
combined = combined.drop(['Team 1', 'Team 2', 'City', 'DayOfWeek', 'TimeOfGame'], axis=1)

## Important Dataset Modifications made after One Hot Encodings
- All categorical attributes represented using One Hot vectors now.
- 1 column from each category dropped as all 0's in that category will represent the dropped value. This helps in dimensionality reduction.
- All original Categorical features are dropped as they are now represented as One Hot vectors.

We split back the train and test

In [57]:
train = combined[combined["train"] == 1]
test = combined[combined["train"] == 0]

In [58]:
train.drop("train", axis = 1, inplace = True)

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [59]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 252 entries, 0 to 251
Data columns (total 69 columns):
AvgWindSpeed                                                        252 non-null float64
AvgHumidity                                                         252 non-null float64
Inn 1 Team 1 NOP R>25,SR>125                                        252 non-null float64
Inn 1 Team 1 NOP R<25, SR>125                                       252 non-null float64
Inn 1 Team 1 Total 4s                                               252 non-null float64
Inn 1 Team 1 Total 6s                                               252 non-null float64
Inn 1 Team 1 Max Strike Rate_ALLBatsmen                             252 non-null float64
Inn 1 Team 2 NoP fast bowlers                                       252 non-null float64
Inn 1 Team 2 NoP Spinners                                           252 non-null float64
Inn 1 Team 2 wickets taken_catches_runout                           252 non-null float64
Inn1 

In [60]:
test.drop("train", axis = 1, inplace = True)

## Next, we perform the split for independent and dependent variables

In [61]:
y_train = train['Winner (team 1=1, team 2=0)']

In [62]:
train.drop("Winner (team 1=1, team 2=0)", axis = 1, inplace = True)

In [63]:
y_test = test['Winner (team 1=1, team 2=0)']

In [64]:
test.drop("Winner (team 1=1, team 2=0)", axis = 1, inplace = True)

## Classification

Multiple algorithms would be tried with our data.

- K-Nearest Neighbors
- Decision Trees
- Random Forest
- Support Vector Machines
- Support Vector Machines (Kernel based)

## Model Evaluation

We evaluate our models based on __Accuracy, Precision, Recall and F1-Score__.

__The predict_proba method gives the probability of each team's winning__

### K-Nearest Neighbors

In [65]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(test)

print("Probability predictions for which team would win each match up in order for the test data")
print(classifier.predict_proba(test))

p, r, f1, _ = precision_recall_fscore_support(y_test, y_pred, average='micro')
acc = accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print("Precision: ", p)
print("Recall: ", r)
print("F1 Score: ", f1)

# Making the Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
cm

Probability predictions for which team would win each match up in order for the test data
[[1.  0. ]
 [1.  0. ]
 [0.4 0.6]
 [0.  1. ]
 [0.  1. ]
 [0.2 0.8]
 [0.2 0.8]
 [0.  1. ]
 [0.6 0.4]
 [0.  1. ]
 [1.  0. ]
 [0.  1. ]
 [0.4 0.6]
 [1.  0. ]
 [0.4 0.6]
 [0.8 0.2]
 [0.8 0.2]
 [0.  1. ]
 [1.  0. ]
 [0.8 0.2]
 [1.  0. ]
 [1.  0. ]
 [0.6 0.4]
 [0.  1. ]
 [0.4 0.6]
 [0.8 0.2]
 [0.2 0.8]
 [0.6 0.4]
 [0.8 0.2]
 [0.  1. ]
 [1.  0. ]
 [1.  0. ]
 [0.4 0.6]
 [1.  0. ]
 [0.2 0.8]
 [0.4 0.6]
 [0.2 0.8]
 [0.  1. ]
 [0.6 0.4]
 [0.8 0.2]
 [0.4 0.6]
 [0.4 0.6]
 [0.6 0.4]
 [1.  0. ]
 [0.6 0.4]
 [0.4 0.6]
 [0.4 0.6]
 [0.  1. ]
 [1.  0. ]
 [0.8 0.2]
 [0.8 0.2]
 [1.  0. ]
 [0.  1. ]
 [1.  0. ]
 [0.4 0.6]
 [0.8 0.2]
 [0.  1. ]
 [1.  0. ]
 [1.  0. ]
 [0.  1. ]
 [0.6 0.4]
 [0.4 0.6]
 [0.4 0.6]
 [0.6 0.4]
 [0.2 0.8]
 [0.6 0.4]
 [0.  1. ]
 [0.8 0.2]
 [0.8 0.2]
 [0.2 0.8]
 [0.4 0.6]
 [1.  0. ]
 [0.2 0.8]
 [0.4 0.6]
 [0.  1. ]
 [0.8 0.2]]
Accuracy:  0.8421052631578947
Precision:  0.8421052631578947
Recall:  0.8

array([[34,  8],
       [ 4, 30]])

### Decision Tree Classifier

In [66]:
# Fitting Decision Tree Classification to the Training set
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(train, y_train)

y_pred = classifier.predict(test)

p, r, f1, _ = precision_recall_fscore_support(y_test, y_pred, average='micro')
acc = accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print("Precision: ", p)
print("Recall: ", r)
print("F1 Score: ", f1)

cm = confusion_matrix(y_test, y_pred)
cm

Accuracy:  0.8026315789473685
Precision:  0.8026315789473685
Recall:  0.8026315789473685
F1 Score:  0.8026315789473685


array([[34,  8],
       [ 7, 27]])

### Random Forest Classifier

In [67]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 25, criterion = 'entropy', random_state = 0)
classifier.fit(train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(test)

p, r, f1, _ = precision_recall_fscore_support(y_test, y_pred, average='micro')
acc = accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print("Precision: ", p)
print("Recall: ", r)
print("F1 Score: ", f1)
# print(classifier.predict_proba(train))

# Making the Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
cm


Accuracy:  0.9078947368421053
Precision:  0.9078947368421053
Recall:  0.9078947368421053
F1 Score:  0.9078947368421053


array([[37,  5],
       [ 2, 32]])

### Support Vector Machines

In [68]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0, probability=True)
classifier.fit(train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(test)


p, r, f1, _ = precision_recall_fscore_support(y_test, y_pred, average='micro')
acc = accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print("Precision: ", p)
print("Recall: ", r)
print("F1 Score: ", f1)

# Making the Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
cm

Accuracy:  0.9078947368421053
Precision:  0.9078947368421053
Recall:  0.9078947368421053
F1 Score:  0.9078947368421053


array([[40,  2],
       [ 5, 29]])

### Support Vector Machines (RBF kernel)

In [69]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0, probability=True)
classifier.fit(train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(test)

# print(classifier.predict_proba(test))

p, r, f1, _ = precision_recall_fscore_support(y_test, y_pred, average='micro')
acc = accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print("Precision: ", p)
print("Recall: ", r)
print("F1 Score: ", f1)

# Making the Confusion Matrix

cm = confusion_matrix(y_test, y_pred)
cm

Accuracy:  0.9078947368421053
Precision:  0.9078947368421053
Recall:  0.9078947368421053
F1 Score:  0.9078947368421053


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


array([[39,  3],
       [ 4, 30]])

# Conclusions

1. We could do proper Data pre-processing based on the attribute categories.
2. We could apply a few Machine Learning Algorithms and predict which team would win provided the match data.
3. We applied prediction of probabilities as to which team will win for the k-NN clssifier. The same can be added for other algorithms as well.

# Additional Improvements

1. We can modularize the pre-processing and prediction part.
2. We can apply GridSearchCV to figure out the best hyper parameters for our data.
3. We can improve our data so that our model doesn't get penalized for not having seen certain data.
4. We can experiment with identifying improtant features using measures such as Principal Component Analysis, Correlation Analysis etc.